### ISSUE TITLE
'safe' filters

### ISSUE THREAD

**Issue by brigadier** [erlydtl/erlydtl#247]
Is it possible to mark result of a custom filter so that it would never get escaped, always 'safe'?

**Comment by kaos** [comment]
I'll have to check..

**Comment by kaos** [comment]
No, I don't think so. Currently, there is an explicit check in `filter_tag_ast` that checks if the filter _name_ is one of `escape`, `safe` or `safeseq` and flags the content safe for those only.

Reading up a bit on the django docs, I think the way to go is to use the `safe` (or `safeseq` for lists) filter after your custom filter..

`... {{ var1 | customFilter | safe }} ... `

**Comment by brigadier** [comment]
In django there's a `mark_safe` function https://docs.djangoproject.com/en/1.10/ref/utils/#django.utils.safestring.mark_safe
When a filter or tag returns `mark_safe("<script>alert(123);</script>")` it won't be escaped.

**Comment by brigadier** [comment]
maybe if filter or tag returns tuple `{safe, Str::iolist())` just output the `Str` to the next filter or in the template without futher escaping?

**Comment by kaos** [comment]
I'm more inclined to adapt the filter/tag api to include a context that should be returned along with the data. That way, it can handle this, and other future needs without too much hassle.

**Comment by kaos** [comment]
@brigadier you're idea was better than mine (much easier to implement at least). It has the side effect of also applying to variables, so `{{ var1 }}` can be flagged safe when `var1 = {safe, Data}`.

Pushed to `develop`. If it works out, and there's not a lot of people crying out that they rely on a `{safe, _}` tuple being output as-is, we can merge it to master in a while..

**Comment by seriyps** [comment]
I'm not sure if it plays well with filter pipelining like `{{ var1 | filter_returning_safe | some_other_filter }}` or such usecase considered as invalid? AFAIK, django uses python's polymorphism: strings, marked by `mark_safe` looks like and can be handled like usual strings, until you implicitly check if it is 'safe string'...

**Comment by kaos** [comment]
@seriyps good point. The context approach I was referring to would work better here, the trouble was that the filter result is render-time whereas currently, the safe decision is compile-time, so that requires some major modifications to get going. 

This will require some more thought to get a good enough solution, then..

**Comment by seriyps** [comment]
I may suggest smth like compile-time 'tag capabilities'. So, compiler in compile-time asks filter's module if it's output is safe and generate different code for safe and unsafe filters.

But we can just accept current proposed solution, since it covers major amount of usecases.

**Comment by kaos** [comment]
> compiler in compile-time asks filter's module if it's output is safe and generate different code for safe and unsafe filters.
I like it! :)

**Comment by manuel-rubio** [comment]
what was finally the better idea? pass the context as parameter to the filters or optionally into the return? about the idea of returning `{safe, String}` I think it's ok, we have to keep in mind only to remove that safe if we are putting other filters before and, of course, safe is lost in that case.
